compiler/src/jdk.graal.compiler/src/jdk/graal/compiler/nodes/GraphDecoder.java

```diff
@@ -1169,9 +1217,43 @@ protected void handleLoopExplosionBegin(MethodScope methodScope, LoopScope loopS
         }
 
         if (methodScope.loopExplosion.mergeLoops()) {
-            LoopExplosionState explosionState = new LoopExplosionState(frameState, merge);
-            loopScope.iterationStates.put(explosionState, explosionState);
+            LoopExplosionKey explosionKey = createLoopExplosionKey(loopScope, frameState);
+            LoopExplosionState explosionState = new LoopExplosionState(frameState, merge, loopScope.loopDepth);
+            methodScope.iterationStates.put(explosionKey, explosionState);
+        }
+    }
+
+    protected LoopExplosionKey createLoopExplosionKey(LoopScope loopScope, FrameState frameState) {
+        List<ValueNode> values;
+        if (loopScope.loopExplosionMergeKeySlots != null) {
+            values = new ArrayList<>(loopScope.loopExplosionMergeKeySlots.size());
+            for (int i : loopScope.loopExplosionMergeKeySlots) {
+                ValueNode node = frameState.values().get(i);
+                if (!node.isConstant() || node.asJavaConstant().getJavaKind() != JavaKind.Int) {
```

Comment:
Should we accept all integer types (i.e., also accept longs)?
What do you think @chumer?

Reply:
Just ints are fine for now imo.